src/data/set/ncard.lean

```diff
@@ -0,0 +1,696 @@
+allows cardinality computations to avoid `finset`/`fintype` completely, staying in `set` and letting
+finiteness be handled explicitly, or (where a `finite α` instance is present and the sets are
+in `set α`) via `auto_param`s.
+
+## Main Definitions
+
+* `set.ncard s` is the cardinality of the set `s` as a natural number, provided `s` is finite.
+  If `s` is infinite, then `set.ncard s = 0`.
+* `to_finite_tac` is a tactic that tries to synthesize an `set.finite s` argument with
+  `set.to_finite`. This will work for `s : set α` where there is a `finite α` instance.
+
+## Implementation Notes
+
+The lemmas in this file are very similar to those in `data.finset.card`, but with `set` operations
+instead of `finset`; most of the proofs invoke their `finset` analogues. Nearly all the lemmas
+require finiteness of one or more of their arguments. We provide this assumption with an
+`auto_param` argument of the form `(hs : s.finite . to_finite_tac)`, where `to_finite_tac` will find
+a `finite s` term in the cases where `s` is a set in a `finite` type.
+
+Often, where there are two set arguments `s` and `t`, the finiteness of one follows from the other
+in the context of the lemma, in which case we only include the ones that are needed, and derive the
+other inside the proof. A few of the lemmas, such as `ncard_union_le` do not require finiteness
+arguments; they are are true by coincidence due to junk values.
+-/
+
+open_locale classical
+
+variables {α β : Type*} {s t : set α} {a b x y : α} {f : α → β}
+
+namespace set
+
+
+/-- The cardinality of `s : set α` . Has the junk value `0` if `s` is infinite -/
+noncomputable def ncard (s : set α) := nat.card s
+
+/-- A tactic that finds a `t.finite` term for a set `t` in a `finite` type. -/
+meta def to_finite_tac : tactic unit := `[exact set.to_finite _]
+
+lemma ncard_def (s : set α) : s.ncard = nat.card s := rfl
+
+lemma ncard_eq_to_finset_card (s : set α) (hs : s.finite . to_finite_tac) :
+  s.ncard = hs.to_finset.card :=
+by rw [ncard_def, @nat.card_eq_fintype_card _ hs.fintype,
+  @finite.card_to_finset _ _ hs.fintype hs]
+
+lemma ncard_le_of_subset (hst : s ⊆ t) (ht : t.finite . to_finite_tac) :
+  s.ncard ≤ t.ncard :=
+@finite.card_le_of_embedding _ _ (finite_coe_iff.mpr ht) (set.embedding_of_subset _ _ hst)
+
+lemma ncard_mono [finite α] :
+  @monotone (set α) _ _ _ ncard :=
+λ _ _, ncard_le_of_subset
+
+@[simp] lemma ncard_eq_zero (hs : s.finite . to_finite_tac) :
+  s.ncard = 0 ↔ s = ∅ :=
+by simp [ncard_def, @finite.card_eq_zero_iff _ hs.to_subtype]
+
+@[simp] lemma ncard_coe_finset (s : finset α) :
+  (s : set α).ncard = s.card :=
+by rw [ncard_eq_to_finset_card, finset.finite_to_set_to_finset]
+
+@[simp] lemma ncard_empty (α : Type*) :
+  (∅ : set α).ncard = 0 :=
+by simp only [ncard_eq_zero]
+
+lemma infinite.ncard (hs : s.infinite) :
+  s.ncard = 0 :=
+@nat.card_eq_zero_of_infinite _ hs.to_subtype
+
+lemma ncard_univ (α : Type*):
```

Comment:
This should be proven from the equivalence between `↥(@univ α)` and `α`. It almost surely exists somewhere in Mathlib, and should be added otherwise.